### ISSUE TITLE
Sensors have problem to import results because of symbolic links and path normalization

### ISSUE THREAD

**Issue by xinglinyin** [SonarOpenCommunity/sonar-cxx#1574]
I use sonarqube 7.3 and sonar-cxx 1.1.0  and cppcheck 1.84

I use cppcheck scan src code result with cppcheck-result-1.xml  eg:
![image](https://user-images.githubusercontent.com/26497032/46796426-47909b00-cd7f-11e8-95bd-5d6e8487f5a6.png)
When i startup sonar-scnaner to parse cppcheck-result-1.xml , raise the issue: 
Cannot find the file 'src/targeting_service_dumpads.cpp', skipping violations

I lookup the official wiki: but can not resolve the problem
my sonar-project.conf such as:
![image](https://user-images.githubusercontent.com/26497032/46796640-b968e480-cd7f-11e8-9f28-fd66d2229449.png)

Then My sonar root path such as:
![image](https://user-images.githubusercontent.com/26497032/46796706-e1f0de80-cd7f-11e8-84a3-6652bc8c2347.png)

any one can help me sovle this issue？  Thank you very much

@guwirth  can help me ?

**Comment by ivangalkin** [comment]
Hi @xinglinyin,

I suggest you to...

1. add property `sonar.projectBaseDir` and set it to absolute directory to your project
2. define `sonar.sources` as a relative pats to the `sonar.projectBaseDir`

... and try to execute the analysis again. Please see https://docs.sonarqube.org/display/SCAN/Advanced+SonarQube+Scanner+Usages

Also there are cases, when seeing this warning is ok. E.g. if you use single [cppcheck] report for multi-module project. Please see my comments in #1508 for more details.

Thank you

**Comment by ivangalkin** [comment]
Hi @xinglinyin,

one more hint regarding to your problem: Previously I suggested you to set the property `sonar.projectBaseDir` to the absolute path of `~/TARGETINGCPP/tmp/` e.g. `/home/braver/tmp` or such. You can skip this property if your working directory (the directory where sonar-scanner is executed from) **IS** you project base directory (the one, which is used in order to resolve all relative paths). If the `sonar.projectBaseDir` was not set, sonar-scanner considers current directory as project base.

```BASH
cd ~/TARGETINGCPP/tmp/
sonar-scanner ...
```

Having a correct project directory is important. All relative paths from external reports are made absolute by means of this project base path.

If the problem persists, please run sonar-scanner in [debug mode](https://github.com/SonarOpenCommunity/sonar-cxx/wiki/Get-debug-information) and upload the log here. Thank you.

**Comment by xinglinyin** [comment]
@ivangalkin  thank you very much firstly.  
I have try your suggests. but don't worked.  the warning still exists.  
I can't not get the correct cppcheck result by soanr-qube.  has no any violatians on sonarqube web . but the cppcheck-result-1.xml like above really has bugs.

and I found that.  when I change the cppcheck report like below. the scanner can parse cppcheck result normally.
![image](https://user-images.githubusercontent.com/26497032/46870595-97e02980-ce61-11e8-9be1-9e6baa795262.png)

so is there any bug abourt soanr-cxx? 
the CppcheckV2 parse the  cppcheck result xml by --xml-version=2 always raise "Cannot find the file 'src/targeting_service_dumpads.cpp', skipping violations"

**Comment by ivangalkin** [comment]
Hi @xinglinyin,

thank you for your investigation! Indeed, there seems to be a problem in parsing of cppcheck XML reports version 2 (whereby the version 1 works well). Unfortunately we have not enough logs in order to trace the root cause.

Could you please upload your project settings (the properties file) as well as the full trace as text file?

In order to address the traceability I've upload a push request. I would appreciate your help very much, if you could try the corresponding patch. Please see https://ci.appveyor.com/project/SonarOpenCommunity/sonar-cxx/builds/19474183/artifacts for the modified plugin version. Maybe you could upload the debug log of the modified code, too. Thank you very much in advance.

Could you please try the new code.


We

**Comment by xinglinyin** [comment]
@ivangalkin  I try debug the sonar-cxx source code. then I found that.  sonar-cxx can not process symbol software link where project root like this:


I guess the the code below has bug 
link url: https://github.com/SonarOpenCommunity/sonar-cxx/blob/bd5adde5c548f6d07cba979036f1a323cd0a8f2c/cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxIssuesReportSensor.java
![image](https://user-images.githubusercontent.com/26497032/46901045-e7b10600-cede-11e8-9f18-8b16df614111.png)

**Comment by ivangalkin** [comment]
Yes, I noticed this questionable logic too. Sonar API has special queries for checking if relative paths (like in your cppcheck report) are part of the project. I addressed this issue in my PR. Do you have a chance to try it?

**Comment by ivangalkin** [comment]
P.S. please see #1575 and (if possible) try https://ci.appveyor.com/project/SonarOpenCommunity/sonar-cxx/builds/19474183/artifacts

Thank you, @xinglinyin

**Comment by ivangalkin** [comment]
@guwirth @all

1. I could reproduce the bug locally: If sensor imports an external report, each file path mentioned in this report becomes checked, whether it's part of SonarQube project (module) or not. If some part of this path is a symbolic link - the check works wrong and return `false`.

The root cause for this problem is that our path normalization (`CxxUtils::normalizePath()`) is implemented through `File::getCanonicalPath()`. This method resolves symbolic links (see https://www.journaldev.com/848/java-file-path-absolute-canonical), which makes the queries like...

```JAVA
String normalPath = CxxUtils.normalizePathFull(path, moduleBaseDir); //  canonical path given
InputFile inputFile = sensorContext.fileSystem().
inputFile(sensorContext.fileSystem().predicates().hasAbsolutePath(normalPath)); // absolute path requested
```

... obviously go wrong.

2. In my PR #1575 I rewrote the snippet above with the query

```JAVA
final InputFile inputFile = sensorContext.fileSystem().inputFile(sensorContext.fileSystem().predicates().hasPath(path));
```
It is designed to work with both relative and absolute `path`s. And indeed it does work well with symbolic links too. I double checked it on my local setup.

3. There are still sensors, which use `CxxUtils.normalizePathFull()`. I will rewrite them through `java.nio.Path::normalize()`. This method doesn't resolves the symbolic links (see https://docs.oracle.com/javase/tutorial/essential/io/pathOps.html#normal)
